### PR TITLE
Adds a module to support gtsummary

### PR DESCRIPTION
**Pull request by llrs-roche**
Fixes https://github.com/insightsengineering/coredev-tasks/issues/714
Related to https://github.com/insightsengineering/coredev-tasks/issues/676

Adds 3 more package dependencies: (methods), crane, gtsummary, gt. I haven't added minimal version for them.
We could remove crane dependency at the cost of having less corporate style tables (instead of using `crane::tbl_roche_summary()` we could have the vanilla `gtsummary::tbl_summary()`.

Documentation is mostly imported from other packages.

It involves just 3 functions: the module (is this a good name for such module?), the ui and the server function.
Support to download the table depends on https://github.com/insightsengineering/teal.widgets/issues/335. 

Test will come once this branch is reviewed, but known corner case from the module: No `include` variables should select all variables except the one used to split the columns `by`.

Further modifications of the table should be done via decorators not included on this PR.
Adding decorators might involve modifying `check_decorators()` to handle more than one decorator for the same object. This function is duplicated on TMG and TMC and I haven't included on the PR but here is the one I was using:

<details>
<summary>Modified check_decorators to work with multiple decorators for the same object</summary>

```r
check_decorators <- function(x, names = NULL) { # nolint: object_name.

  check_message <- checkmate::check_list(x, names = "named")

  if (!is.null(names) && isTRUE(check_message)) {
    if (length(names(x)) != length(unique(names(x)))) {
      check_message <- sprintf(
        "The `decorators` must contain unique names from these names: %s",
        paste(names, collapse = ", ")
      )
    }
  } else {
    check_message <- sprintf(
      "The `decorators` must be a named list from these names: %s",
      paste(names, collapse = ", ")
    )
  }

  if (!isTRUE(check_message)) {
    return(check_message)
  }

  valid_elements <- vapply(
    x,
    checkmate::test_class,
    classes = "teal_transform_module",
    FUN.VALUE = logical(1L)
  )

  # Nested list
  if (any(!valid_elements)) {
    valid_nested <- vapply(
      x[!valid_elements], function(subdecorators) {
        checks <- vapply(subdecorators,
               checkmate::test_class,
               classes = "teal_transform_module",
               logical(1L))
        all(checks)
      },
      FUN.VALUE = logical(1L)
    )
    if (all(valid_nested)) {
      return(TRUE)
    }
  }

  if (all(valid_elements)) {
    return(TRUE)
  }

  "Make sure that the named list contains 'teal_transform_module' objects created using `teal_transform_module()`"
}
```

</details>

<details>
<summary>Example:</summary>

```r
devtools::load_all()
data <- within(teal.data::teal_data(), {
    ADSL <- teal.data::rADSL
})
join_keys(data) <- default_cdisc_join_keys[names(data)]
app <- init(
    data = data,
    modules = modules(
        tm_gt_summary(
            by = teal.transform::data_extract_spec(
                dataname = "ADSL",
                select = teal.transform::select_spec(
                    choices = c("SEX", "COUNTRY", "SITEID", "ACTARM"),
                    selected = "SEX",
                    multiple = FALSE
                )
            ),
            include = teal.transform::data_extract_spec(
                dataname = "ADSL",
                select = teal.transform::select_spec(
                    choices = c("SITEID", "COUNTRY", "ACTARM"),
                    selected = "SITEID",
                    multiple = TRUE,
                    fixed = FALSE
                )
            )
        )
    )
)
if (interactive()) {
    shinyApp(app$ui, app$server)
}

```

</details>

### CODE REVIEW

File: R/gtsummary.R [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+#' )
+#' if (interactive()) {
+#'   shinyApp(app$ui, app$server)
+#' }
+tm_gt_summary <- function(
+  label = "Table summary",
+  by,
+  ...,
+  col_label = NULL,
+  include,
```
Comment (osenan) [inline R/gtsummary.R]: I think include should be before the `...` as is [a mandatory argument](https://tidydesign.substack.com/p/argument-ordering).

File: R/gtsummary.R [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+      any.missing = FALSE, all.missing = FALSE
+    )
+    assert_single_selection(list(by))
+  }
+  if (inherits(include, "data_extract_spec")) {
+    checkmate::assert_list(list(include),
+      types = "data_extract_spec",
+      any.missing = FALSE, all.missing = FALSE
+    )
+  }
```
Comment (osenan) [inline R/gtsummary.R]: Let's add assertion on `pre_output` and `post_output` as we will use them in the ui
  
```suggestion
  }
checkmate::assert_multi_class(pre_output, c("shiny.tag", "shiny.tag.list", "html"), null.ok = TRUE)
checkmate::assert_multi_class(post_output, c("shiny.tag", "shiny.tag.list", "html"), null.ok = TRUE)
```

File: R/gtsummary.R (hunk 1 above)
Comment (osenan) [reply R/gtsummary.R]: Now, looking at the server code I understand. User should provide `by` or `include` arguments. I think `by` is mandatory and `include` is not. Please just confirm if that is the case

**Review by osenan — CHANGES_REQUESTED** [review]
Hi, very good job!

I have provided small comments on function arguments and assertions. I have reviewed the full ui part and all parts of the server that I was confident. The last elements when we "print" the content of the qenv to subsequent qenv looks slightly unknown to me.

I cannot see the table in the example you provided, is it working for you?

Image

On new iteration I will check the documentation.

File: DESCRIPTION [hunk 3]
```diff
@@ -43,11 +44,14 @@ Imports:
...
+    gt,
+    gtsummary,
     htmlwidgets (>= 1.6.4),
     jsonlite (>= 1.8.9),
     lattice (>= 0.18-4),
     lifecycle (>= 0.2.0),
     MASS (>= 7.3-60),
+    methods,
     rmarkdown (>= 2.23),
     rtables (>= 0.6.11),
```
[the remark this answers is not in the document]
Comment (llrs-roche) [reply DESCRIPTION]: Other modules still use rtables until we remove them we still depend on it.

File: R/gtsummary.R
[the remark this answers is not in the document]
Comment (llrs-roche) [reply R/gtsummary.R]: Yes, my oversight, I will rename the file once we settle on a name for the module

**Review by osenan — APPROVED** [review]
Thanks for the changes. I have now tested the app. All the `validate::need` calls are appearing. Let's find more bugs with more dry runs or with the unit tests let's check the edge cases.

I found this curiosity of calling a hash function on the report:
```
ADSL <- teal.data::rADSL
stopifnot(rlang::hash(ADSL) == "843e317c3d4aeb88062cd39a9c62fe8a") # @linksto ADSL
.raw_data <- list2env(list(ADSL = ADSL))
lockEnvironment(.raw_data) # @linksto .raw_data
library(crane)
table <- crane::tbl_roche_summary(data = ADSL, by = "SITEID", include = c("SITEID", "ACTARM"), nonmissing = "ifany", percent = "column")
table
```

Just decide if you want to move the include before the ... or you prefer to keep it where it is. In that case, let's see if it can be set to `NULL`

**Comment by llrs-roche** [comment]
Waiting to merge to see if we can merge a feature complete module with the work on:

- teal.widgets: https://github.com/insightsengineering/teal.widgets/pull/337
- tests: https://github.com/insightsengineering/teal.modules.general/pull/975

File: inst/WORDLIST [hunk 4]
```diff
@@ -16,3 +16,4 @@ qq
 reportable
 sortable
 tabset
+gtsummary
```
Comment (m7pr) [inline inst/WORDLIST]: I would remove from WORDLIST, and would write gtsummary with ticks all over the documentation

File: R/tm_gtsummary.R [hunk 5]
```diff
@@ -0,0 +1,313 @@
+#' `teal` module: Summary table
```
Comment (m7pr) [inline R/tm_gtsummary.R]:
```suggestion
#' `teal` module: GT Summary table
```

File: R/tm_gtsummary.R [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+#'
+#' Generates a table summary from a dataset using gtsummary.
+#'
+#' @inheritParams teal::module
+#' @inheritParams shared_params
+#' @param by (`data_extract_spec` or `list` of multiple `data_extract_spec`)
+#' Object with all available choices with pre-selected option for how to split the rows
```
Comment (m7pr) [inline R/tm_gtsummary.R]:
```suggestion
#' An object with all available choices and with a pre-selected option on how to split rows.
```

**Review by m7pr — COMMENTED** [review]
For now reviewed documentation and DESCRIPTION

File: R/tm_gtsummary.R [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+#'           fixed = FALSE
+#'         )
+#'       )
+#'     )
+#'   )
+#' )
+#' if (interactive()) {
+#'   shinyApp(app$ui, app$server)
+#' }
+tm_gtsummary <- function(
```
[the remark this answers is not in the document]
Comment (m7pr) [reply R/tm_gtsummary.R]: Also added this comment on Slack

**Comment by m7pr** [comment]
So far I was able to run the example from the opening of this PR on a `tests` branch.
And I was able to create a table, add it to the report and display in a report

image

Will now try to test with a decorator and transformator as well

File: R/tm_gtsummary.R [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+  label = "Table summary",
```
[the remark this answers is not in the document]
Comment (llrs-roche) [reply R/tm_gtsummary.R]: Thanks! 🤦‍♂️
